### PR TITLE
[Feature] Device transform

### PR DESCRIPTION
**Pull request by vmoens**
## Description

Adds a `DeviceCastTransform` transform to move environment data from one device to another.

As part of this PR, transforms now can transform the device of the parent env through `transform.transform_env_device`.

### CODE REVIEW

File: torchrl/envs/transforms/transforms.py [hunk 1]
```diff
@@ -2708,6 +2738,87 @@ def __init__(
...
+        return tensordict.to(self.device, non_blocking=True)
+
+    def set_container(self, container: Union[Transform, EnvBase]) -> None:
+        if self.orig_device is None:
+            if isinstance(container, EnvBase):
+                self.orig_device = container.device
+                container._device = self.device
+            elif container.parent is not None:
+                self.orig_device = container.parent.device
+                raise RuntimeError("How can we change the device of the container?")
```
Comment (matteobettini) [inline torchrl/envs/transforms/transforms.py:2788]: Is this supposed to be here?

File: torchrl/envs/transforms/transforms.py [hunk 2]
```diff
@@ -2708,6 +2738,87 @@ def __init__(
         super().__init__(torch.double, torch.float, in_keys, in_keys_inv)
 
 
+class DeviceCastTransform(Transform):
+    """Casts the env device.
```
Comment (matteobettini) [inline torchrl/envs/transforms/transforms.py:2742]: Is this really casting the env device?

Isn't it transforming the device of the data?

**Review by matteobettini — APPROVED** [review]
I have a question:

In environments like VMAS where there is an internal state that sits on a device, the way I thought that the environment is moved to another device is `env.to(device)` 

If i understand correctly here we are casting the env data and its specs to another device, but not its internal state (we do not call `env.to()`). 

does this make sense to use with a parent environment? when is this transform preferred to `env.to(device)`?

the only case i can imagine is when we need to keep an env of a specific device, apply some transforms there and then move the data to another device

**Comment by vmoens** [comment]
> does this make sense to use with a parent environment? when is this transform preferred to `env.to(device)`?

This is to address https://github.com/pytorch/rl/issues/1198 where the issue is that if the env naturally sits on MPS we can't use float64. So first you must transform the data into float32 and then cast to device. Doing env.to(device) will not work but this transform will.

**Comment by EkaterinaAbramova** [comment]
> > does this make sense to use with a parent environment? when is this transform preferred to `env.to(device)`?
> 
> This is to address #1198 where the issue is that if the env naturally sits on MPS we can't use float64. So first you must transform the data into float32 and then cast to device. Doing env.to(device) will not work but this transform will.

Could you please explain exactly how to do this? I am confused. I am flooding this tutorial https://pytorch.org/tutorials/intermediate/reinforcement_ppo.html and get that error about MPS flaot64 on the line: base_env = GymEnv("InvertedDoublePendulum-v4", device=device, frame_skip=frame_skip) What code exactly shall I write to correct this error please?

**Comment by vmoens** [comment]
Can I ask what the value of device is in your case?

**Comment by EkaterinaAbramova** [comment]
device="mps". 

I believe I solved it with this (after quite a few hours of trying different things!!!):

```
base_env = GymEnv("InvertedDoublePendulum-v4", device="cpu", frame_skip=frame_skip)
env = TransformedEnv(
    base_env,
    Compose(
        ObservationNorm(in_keys=["observation"]), # normalise observations (make it about Standard Normal)
        DoubleToFloat(),   
        StepCounter(),                            # count the number of steps before the environment is terminated
        DeviceCastTransform(device=device, orig_device="cpu"),
    ),
)
print(env.device) # gives mps now
```

Could you please kindly confirm if what Ive done is correct? I am on Apple M2 max trying to use MPS.

**Comment by EkaterinaAbramova** [comment]
Everything was progressing smoothly through the tutorial: https://pytorch.org/tutorials/intermediate/reinforcement_ppo.html however at this code I get an error again about MPS. Please could you kindly advise syntax to solve this issue? I double checked everything seems to be on mps, so I don't understand where the error is coming from.

```
collector = SyncDataCollector(
    env,
    policy_module,
    frames_per_batch=frames_per_batch,
    total_frames=total_frames,
    split_trajs=False,
    device=device,
)


  File ~/anaconda3/envs/gpu-torch-rl/lib/python3.10/site-packages/torch/nn/modules/module.py:1143 in convert
    return t.to(device, dtype if t.is_floating_point() or t.is_complex() else None, non_blocking)

TypeError: Cannot convert a MPS Tensor to float64 dtype as the MPS framework doesn't support float64. Please use float32 instead.


env.device
Out[23]: device(type='mps')
```
Screenshot 2023-09-30 at 01 36 12

Screenshot 2023-09-30 at 01 38 39

**Comment by vmoens** [comment]
@EkaterinaAbramova sorry you had this terrible experience, we should document things better for MPS. 
It's something we're actively looking at and any feedback is very welcome.
The way you implemented your env looks great to me!

Regarding the error I will have a look into it, shouldn't be too difficult to solve. Have you tried moving the `ObservationNorm` and `StepCounter` after the device casting transform? 
like:
```python
env = TransformedEnv(
    base_env,
    Compose(
        DoubleToFloat(),   
        DeviceCastTransform(device=device, orig_device="cpu"),
        ObservationNorm(in_keys=["observation"]), # normalise observations (make it about Standard Normal)
        StepCounter(),                            # count the number of steps before the environment is terminated
    ),
)
```
Like this the buffers in the `ObservationNorm` transform will sit on mps but with float32 and not float64.

**Comment by EkaterinaAbramova** [comment]
@vmoens thank you for swiftly helping me, this issue is quite urgent, so very glad that you had the suggestion. It makes sense and I tried it, however this way around I get this error `AttributeError: 'DoubleToFloat' object has no attribute 'init_stats'` (it seems I need to pass some arguments, what shall I pass to be able to follow the tutorial please https://pytorch.org/tutorials/intermediate/reinforcement_ppo.html): 

Screenshot 2023-09-30 at 23 26 42

**Comment by vmoens** [comment]
You need to call init_stats on the obs norm transform
```python
env.transform[-2].init_stats(...)
```
Because the transform has changed place

**Comment by EkaterinaAbramova** [comment]
OK I get it. So now that Ive indexed the correct location, the MPS issue is back. 
The code you suggested doesn't work for me:
```
env = TransformedEnv(
    base_env,
    Compose(
        DoubleToFloat(),   
        DeviceCastTransform(device=device, orig_device="cpu"),
        ObservationNorm(in_keys=["observation"]), # normalise observations (make it about Standard Normal)
        StepCounter(),                            # count the number of steps before the environment is terminated
    ),
)
print(env.device)
env.transform[-2].init_stats(num_iter=1000, reduce_dim=0, cat_dim=0) 
I get the error: Cannot convert a MPS Tensor
```

TO RECAP: The original code was: 
```
env = TransformedEnv(
    base_env,
    Compose(
        ObservationNorm(in_keys=["observation"]), # normalise observations (make it about Standard Normal)
        DoubleToFloat(),   
        StepCounter(),                            # count the number of steps before the environment is terminated
        DeviceCastTransform(device=device, orig_device="cpu"),
    ),
)
env.transform[0].init_stats(num_iter=1000, reduce_dim=0, cat_dim=0) 
```
This code runs, however I have issues further below in the tutorial where I can't run:
```
collector = SyncDataCollector(
    env,
    policy_module,
    frames_per_batch=frames_per_batch,
    total_frames=total_frames,
    split_trajs=False,
    device=device,
)
Cannot convert a MPS Tensor
```

Could you please provide a solution? I am really in a rush now, Ive been trying to solve this issue for days. The tutorial I am working on is online, so if it helps maybe you could try the suggestions to make sure they resolve the issue? Thank you very much for your help, I really need to move past this ASAP please.

**Comment by vmoens** [comment]
Ok so that's an interesting bug, which basically boils down to some internal machinery within rollout, resets and transforms.
To quickly unblock you: can you compute the stats manually with your env?
```python
simple_env = TransformedEnv(
    base_env,
    Compose(
        DoubleToFloat(),   
        DeviceCastTransform(device=device, orig_device="cpu"),
   )
)
td0 = simple_env.rollout(100)
loc = td0["observation"].mean(dim=0)
scale = td0["observation"].std(dim=0)
env = TransformedEnv(
    base_env,
    Compose(
        DoubleToFloat(),   
        DeviceCastTransform(device=device, orig_device="cpu"),
        ObservationNorm(in_keys=["observation"], loc=loc, scale=scale),
        StepCounter(),
    ),
)
```
Hopefully that should help!
I should be able to put my hands on an apple silicon computer tomorrow morning if you're still stuck!

**Comment by EkaterinaAbramova** [comment]
The suggestion didn't work Im afraid. Any other thing you could propose at this stage or only tomorrow? 

Screenshot 2023-10-01 at 21 27 05

So why in the version I provided above, everything is fine until I get down to the SyncDataCollector? Why is it failing there? What has not been yet converted to float32?

**Comment by vmoens** [comment]
#1589 will solve your problem!

**Comment by vmoens** [comment]
@EkaterinaAbramova you should be good to go now!

**Comment by vmoens** [comment]
I can reproduce your issue, let me push a fix!

**Pull request merged**